rust/lance/src/dataset.rs

```diff
@@ -486,17 +488,63 @@ impl Dataset {
             });
         }
 
-        let fragments = write_fragments(
-            object_store.clone(),
-            &self.base,
-            &schema,
-            stream,
-            params.clone(),
-        )
-        .await?;
+        let transaction = match (params.tail_compaction_size, self.get_fragments().last()) {
+            (Some(tail_compact_thresh), Some(last_frag))
+                if last_frag.count_rows().await? <= tail_compact_thresh =>
+            {
+                let last_data = last_frag.scan().try_into_stream().await?;
+                let last_frag_stream: SendableRecordBatchStream = last_data.into();
+                let arrow_schema = last_frag_stream.schema();
+                let merged_stream = {
+                    // need to scope this use because TokioStreamExt collides with futures::StreamExt
+                    use tokio_stream::StreamExt as TokioStreamExt;
+                    RecordBatchStreamAdapter::new(arrow_schema, last_frag_stream.merge(stream))
+                };
 
-        let transaction =
-            Transaction::new(self.manifest.version, Operation::Append { fragments }, None);
+                let fragments = write_fragments(
+                    object_store.clone(),
+                    &self.base,
+                    &schema,
+                    Box::pin(merged_stream),
+                    params.clone(),
+                )
+                .await?;
+
+                if fragments.len() > 1 {
+                    return Err(Error::InvalidInput {
+                        source: "When attempting to run tail compaction, ".into(),
+                    });
+                }
+                // maybe we should roll append into a rewrite somehow?
+                // append and rewrite are both remove and/or appending fragments
+
+                // Concurrent writes fails on this path
+                // NOCOMMIT: explain how this interacts with concurrent
+                Transaction::new(
+                    self.manifest.version,
+                    Operation::Rewrite {
```

Comment:
We should really create a new operation for this. The semantics of `Rewrite` are meant to be "I only moved data; I didn't modify or add or remove any data". This has implications in conflict resolution.